Kaptive v2.0.0

stitam opened the pull request:
Hi,

Many thanks for maintaining this repo and the docker containers, they are super useful!

- [x] This comment contains a description of what is in the pull request.

This pull request contains the docker file and other updates for Kaptive v2.0.0:
https://www.biorxiv.org/content/10.1101/2021.11.05.467534v2.abstract
https://github.com/katholt/Kaptive

<!-- If this PR is to add a dockerfile for a new tool -->
- [x] Build your own docker image using a Dockerfile
  - [x] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [x] Includes the recommended LABELS
- [x] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/) - solved, apart from adding version numbers to linux dependencies
- [x] Edit main README.md
- [x] Edit Program_Licenses.md
- [x] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [x] Write a GitHub actions workflow
  - [x] Should be located in .github/workflows/ and named test-<tool>.yml (i.e. .github/workflows/test-spades.yml)
  - [x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
  - [x] Have successfully run the workflow "Test <program name> image" in your forked repository

GitHub Actions returns the following error - If you could help me sort this out that would be great!
`buildx failed with: error: failed to solve: target stage test could not be found`

The container works as intended: https://hub.docker.com/r/stitam/kaptive

stitam commented:
I have just realised v2.0.2 was released yesterday so I'm going to push that version as well before flagging the PR ready for review.

kapsakcj commented:
Hi Tamás, thanks for the PR! We have Kleborate but not Kaptive as a standalone docker image so thanks for starting a PR. We would love to add it here.

To answer your question about the error message from the GH actions workflow - you'll need to have two separate `FROM` layers in your dockerfile. The first being the `app` layer, which will contain everything that is currently in your dockerfile. You can change the first line to `FROM ubuntu:focal as app`

The second FROM layer should look like `FROM app as test` which will designate the testing layer. This testing layer is used for just testing and will not be part of the final docker image when deployed to dockerhub and quay.

Could you please add a test layer to the dockerfile? That should allow the test GH actions workflow to run.

You can see some other more recent dockerfiles as an example, and our guide has some more in-depth info on this.
https://github.com/StaPH-B/docker-builds/blob/master/spades/3.15.4/Dockerfile
https://github.com/StaPH-B/docker-builds/blob/master/ncbi-amrfinderplus/3.10.24/Dockerfile

guide: https://staphb.org/docker-builds/make_containers/

stitam commented:
Thanks @kapsakcj, will do in a moment. Regarding Kleborate: while Kleborate includes Kaptive, the tool is focused on K. pneumoniae, and if I remember well I couldn't get it to work with A.baumannii. On the other hand, Kaptive works well with A. baumannii.

stitam commented:
I am wondering maybe it would be better to merge v2.0.0 for now? I can open another PR after fixing v2.0.2.

stitam commented:
I've added files for Kaptive v2.0.3 (relased a few hours ago) and tested the GitHub workflow for both versions ([v2.0.0](https://github.com/stitam/docker-builds/actions/workflows/test-kaptive-2.0.0.yml), [v2.0.3](https://github.com/stitam/docker-builds/actions/workflows/test-kaptive-2.0.3.yml)).

erinyoung commented on kaptive/2.0.3/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+WORKDIR /data
+
+# test layer
+FROM app as test
```
This one is also missing a test in the test layer.

kapsakcj commented:
I've made empty repos on dockerhub and quay in prep for these docker images:
https://hub.docker.com/r/staphb/kaptive
https://quay.io/repository/staphb/kaptive?tab=info

stitam commented:
Many thanks @erinyoung for reviewing the PR!

I added myself as author/maintainer and added tests to the test layers. I rebuilt the containers, the containers work well on my system and the workflows on my fork complete successfully.

Please note, running the _K. pneumoniae_ K locus prediction on v2.0.0 produces the prediction but also prints some warnings. The new version v2.0.3 also produces the same predictions but without the warnings. It seems this was a Kaptive issue and the Kaptive developers fixed this between v2.0.0 and v2.0.3.

The reference databases which Kaptive uses are quite small, they are part of the [Kaptive GitHub repository](https://github.com/katholt/Kaptive/tree/master/reference_database) and are also inside the container. For example, "/kaptive/reference_database/Klebsiella_k_locus_primary_reference.gbk" is the reference database used for klebsi k locus predictions, and this path is a path within the container.

I usually use kaptive inside a singularity container like this:
```
singularity exec kaptive_2.0.3.sif kaptive.py -a GCA_022268055.1_PDT000434809.1_genomic.fna -k /kaptive/reference_database/Klebsiella_k_locus_primary_reference.gbk
```

Or more like this, suppress some outputs and provide a custom name for the output file:
```
singularity exec kaptive_2.0.3.sif kaptive.py -a GCA_022268055.1_PDT000434809.1_genomic.fna -k /kaptive/reference_database/Klebsiella_k_locus_primary_reference.gbk --no_seq_out --no_json -o "GCA_022268055.1_k_locus"
```

Let me know if I should add anything else to the PR.

erinyoung commented:
Looks good! I fixed a file conflict, which allowed the tests to begin. Once those complete, this will likely get merged and pushed to quay and dockerhub.
